### ISSUE TITLE
ByMeta does not accept `services` as input

### ISSUE THREAD

**Issue by jfcantu** [hashicorp/consul-template#1348]
TL;DR - `services` output cannot be passed to `ByMeta`, even though the documentation says this is valid.

### Consul Template version

consul-template v0.24.1 (58aa6c60)

### Configuration

```liquid
{{- range $subdomain, $services := (services | byMeta "ProxySubdomain") }}
```

### Expected behavior

Per [the documentation](https://github.com/hashicorp/consul-template#bymeta):

> Takes a list of services returned by the `service` or `services` and returns a map that groups services by ServiceMeta values.

### Actual behavior

`2020/02/25 22:04:55.961357 [ERR] (cli) /pomerium/pomerium.yaml.tpl: execute: template: :3:53: executing "" at <"ProxySubdomain">: wrong type for value; expected []*dependency.HealthService; got []*dependency.CatalogSnippet`

### Steps to reproduce

1. Try and pass the output of `services` to `ByMeta`.

### References

It appears (to my amateur eye) that `ByMeta` explicitly expects to be passed a `[]*dep.HealthService`: https://github.com/hashicorp/consul-template/blob/7103eb34608d642d0c7dd3f6d8c2053e166927e6/template/funcs.go#L384

By comparison, `ByTag` accepts an `interface{}`: https://github.com/hashicorp/consul-template/blob/7103eb34608d642d0c7dd3f6d8c2053e166927e6/template/funcs.go#L658

This matches the functions in question - `service` returns a `[]*dep.HealthService` and `services` returns a `[]*dep.CatalogSnippet`:

https://github.com/hashicorp/consul-template/blob/7103eb34608d642d0c7dd3f6d8c2053e166927e6/template/funcs.go#L429

https://github.com/hashicorp/consul-template/blob/7103eb34608d642d0c7dd3f6d8c2053e166927e6/template/funcs.go#L455

**Comment by bobvanderlinden** [comment]
I ran into the same issue. It's very hard to determine what combinations of methods are compatible.

**Comment by eikenb** [comment]
Thanks for reporting this @jfcantu. I replicated it easily and will look into fixing it.

If you or @bobvanderlinden notice any other functions that have an issue like this please report them. I'd prefer people not to have to keep some sort of mental map about what works with what. Thanks.

**Comment by bobvanderlinden** [comment]
@eikenb Thanks! I've found another issue that could be related to the other problem I was facing, so I commented there: https://github.com/hashicorp/consul-template/issues/33

**Comment by jfcantu** [comment]
@eikenb, I'll take a swing at it if you haven't already started.

**Comment by eikenb** [comment]
Found why this wasn't added. The catalog/services API doesn't return the meta-data so you can't group on it. Only the service names and tags. 

It looks like it could be possible by doing a separate catalog/services call for each metadata field as it is available as a filter when calling catalog/services, just not in the return data. I'm trying to see how I'd pass the data from the templates to where the catalog/services call is made as they are separated by a generic interface which makes it interesting.

**Comment by eikenb** [comment]
@jfcantu, please take a look/swing and let me know if you see a good way to do it. Given how byMeta is implemented, as a post call filter, I'm not sure it can be done. Passing the filtered meta-data, as I mentioned above, won't work as it is working on the results which don't contain any metadata.

**Comment by eikenb** [comment]
Also I was wrong about passing in the meta-data to the catalog/services call... it only takes node metadata, not service metadata.

I double check with others internally and we don't have an API call that provides what is needed as the long poll/watch overhead of triggering many of those with lots of services could cause problems.

At this point I'm thinking we might need to just call this a documentation bug and remove the mention of `services` from the byMeta description.

**Comment by eikenb** [comment]
For reference. This is the only API for getting the list of all services.

https://www.consul.io/api/catalog.html#list-services

**Comment by jfcantu** [comment]
@eikenb I'll take your word for it - I'm a relative newbie and had assumed it wouldn't be too much of a change to make it behave more like ByTag, but it sounds like it's more complicated than that.

**Comment by eikenb** [comment]
@jfcantu  My initial gut reaction was the same, that it must have been an oversight and wouldn't be to hard... but turns out that isn't the case.

I'm going to switch this as I indicated (to a docs bug) and fix that. If you'd like this feature to be supported eventually I'd suggest submitting a new feature-request issue for it. We can then keep that in mind moving forward and look for ways to implement it without the killer overhead.

**Comment by eikenb** [comment]
@bobvanderlinden
> I ran into the same issue. It's very hard to determine what combinations of methods are compatible.

 I'm going to file a separate documentation issue about this. Calling out that some sort of table or standard notation is needed to tell what helper functions work with which api functions.

**Comment by bobvanderlinden** [comment]
That already helps quite a lot. Thank you!

**Comment by sodabrew** [comment]
Per my comment https://github.com/hashicorp/consul-template/pull/1120#issuecomment-675775035 so then what _would_ be a working example for fetching service metadata? Is there ever a `.ServiceMeta` or `.Meta` map available, and if so on what object? Or would those only become available if a service is piped through `byMeta`?

**Comment by jfcantu** [comment]
> Per my comment [#1120 (comment)](https://github.com/hashicorp/consul-template/pull/1120#issuecomment-675775035) so then what _would_ be a working example for fetching service metadata? Is there ever a `.ServiceMeta` or `.Meta` map available, and if so on what object? Or would those only become available if a service is piped through `byMeta`?

Buckle up, here we go:

1. `range` over the list of service names returned from `services`.
2. For each element, call `service` on the service name to get a list of service instances.
3. From here you can:
    1. `range` over the list of service instances, or
    2. Just use element 0 if your service instances are identical.
4. Use the `ServiceMeta` field of the service instance.

Example:
```
{{ range services }}
    {{ service . }}
        {{ with index . 0 }}
            {{ index .ServiceMeta "apibasepath" }}
        {{ end }}
    {{ end }}
{{ end }}
```

Internally, `services` returns a list of `CatalogSnippet` objects, which only contain a `Name` and a `Tags` field - `service` returns a `HealthService` object, with all the fields you'd expect.




The documentation isn't as clear as it could be, but basically, you `range` over `services`, and then call `service` on each element of the range, e.g.:

```
{{ range services }}
    {{ with service . }}
        {{ index .ServiceMeta "apibasepath" }}
    {{ end }}
{{ end }}
```

**Comment by sodabrew** [comment]
Thank you @jfcantu! Aha, so what @eikenb wrote in https://github.com/hashicorp/consul-template/issues/1348#issuecomment-595966580
> Found why this wasn't added. The catalog/services API doesn't return the meta-data so you can't group on it. Only the service names and tags.

Is that the return from `/v1/catalog/services` looks like this, just _service names and tags_ as you wrote:

```
$ curl http://127.0.0.1:8500/v1/catalog/services
{
    "consul": [],
    "consul-8300": [],
    "consul-8301": [
        "udp"
    ],
    "consul-8302": [
        "udp"
    ],
    "consul-8500": [],
    "consul-8600": [
        "udp"
    ],
    "counting": [
        "public-api"
    ]
}
```

(The service I'm working on tagging + metadata-ing locally is `hashicorp/counting-service` from the consul docs)
